process.ipynb

```diff
@@ -639,21 +628,32 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "def get_dataframe_from_request(request):\n",
+    "def get_dataframe_from_request(symbol, api_key, session):\n",
     "    \"\"\"\n",
     "    From FRED request's response, get the data as pandas DataFrame.\n",
     "    :return: pandas DataFrame with the values correctly parsed.\n",
     "    \"\"\"\n",
-    "    response = request.json()\n",
-    "    if 'observations' not in response:\n",
-    "        return None\n",
-    "        \n",
-    "    full_response = pd.DataFrame(response['observations'])\n",
-    "    data = full_response.loc[:, ['date', 'value']]\n",
-    "    data['value'] = pd.to_numeric(full_response.value, errors='coerce')\n",
-    "    data = data.set_index(pd.to_datetime(data.date, errors='coerce'), drop=True)\n",
-    "    data.drop('date', axis=1, inplace=True)\n",
-    "    return data.dropna()"
+    "    for retries in range(5):\n",
+    "        try:\n",
+    "            response = download_fred_symbol(symbol, api_key, session).json()\n",
+    "            \n",
+    "            if 'observations' not in response:\n",
+    "                return None\n",
+    "\n",
+    "            full_response = pd.DataFrame(response['observations'])\n",
+    "            data = full_response.loc[:, ['date', 'value']]\n",
+    "            data['value'] = pd.to_numeric(full_response.value, errors='coerce')\n",
+    "            data = data.set_index(pd.to_datetime(data.date, errors='coerce'), drop=True)\n",
+    "            data.drop('date', axis=1, inplace=True)\n",
+    "            \n",
+    "            return data.dropna()\n",
+    "        except Exception as e:\n",
+    "            ex = e\n",
+    "            print(f'Failed to convert JSON data to DataFrame from URL: {url} - retrying')\n",
+    "            sleep(1)\n",
```

Comment:
Thinking if it should raise the exception after the retry else it wont fail?

Reply:
Usually it's only one data source that causes the whole thing to fail, but yes we can make it throw :+1: